### PR TITLE
feat(pdfjs): upgrade to v4 - without breaking backwards compatibility

### PR DESCRIPTION
**Pull request by Bengejd**
This is a continuation of PR #1105 which seems by all accounts to have been abandoned by the original author (@Jordan-Hall). This PR implements all of the requested changes that @shamoon proposed, as well as the original PR implementation of upgrading pdfjs to v4.

## Notable PR Differences:

- pdf.js was updated to use `Promise.withResolvers`, a feature introduced relatively recently. The original PR #1105, addressed this issue by upgrading Angular to version 17. However, I felt this approach was unnecessary, as the only blocker requiring the upgrade was the absence of `Promise.withResolvers`. Instead, I chose to directly polyfill the function by referencing the official [proposal polyfill](https://github.com/tc39/proposal-promise-with-resolvers/blob/main/polyfills.js) and comments from other [PDF-related](https://github.com/wojtekmaj/react-pdf/issues/1811#issuecomment-2157866061) repositories. 

____

If merged, this should close the following:

#### Pull Requests
- #1112 (Low effort duplicate)
- #1105 (Original PR)

### Issues
- #1119 
- #1093 
- #1078

### CODE REVIEW

**Comment by joh-klein** [comment]
Very nice! Thank you so much for picking this up!

File: src/app/pdf-viewer/pdf-viewer.component.ts
Comment (VadimDez) [inline src/app/pdf-viewer/pdf-viewer.component.ts:45]: Please combine the ifs and keep only one comment

**Review by shamoon — APPROVED** [review]
Love how much more focused this PR is and backwards compatiblilty. Thanks for integrating my changes.

Code is pretty straightforward so other than comment from VadimDez, as they say, LGTM. Again, Im not a maintainer tho so defer to him.

Screenshot 2024-08-29 at 1 23 32 PM

File: src/app/pdf-viewer/pdf-viewer.component.ts
Comment (Bengejd) [reply src/app/pdf-viewer/pdf-viewer.component.ts]: @VadimDez I consolidated the ifs statements. The comments on the other hand are a different story... because we're doing an assignment in one line and a type check in another line, Typescript will flag either reference to `Promise.withResolvers` as being not defined in the `PromiseConstructor` (because this is a pollyfill and we need to suppress both messages one way or another). So if you don't want there to be two `@ts-expect-error` statements, I could replace one of them with a `@ts-ignore`, which will still be self-documenting, but less likely to be confused as a random duplicate comment.

**Comment by DerekLiang** [comment]
It looks great. Can we merge this? Thank you!

**Comment by joh-klein** [comment]
Finally!!! 🥳

**Comment by Bengejd** [comment]
Thanks everyone who helped make this a reality! See you all next time a critical security flaw gets found 😉

**Pull request merged**
